src/Swarm/Web.hs

```diff
@@ -93,6 +106,18 @@ mkApp appStateRef =
   goalsHandler = do
     appState <- liftIO (readIORef appStateRef)
     return $ appState ^. gameState . winCondition
+
+  codeRenderHandler contents = do
+    return $ T.pack $ case processTermEither contents of
+      Right (ProcessedTerm (Module stx@(Syntax' _srcLoc _term _) _) _ _) ->
+        drawTree . fmap prettyString . para Node $ stx
+      Left x -> x
+
+  codeRunHandler contents = do
+    foo <- liftIO $ atomicModifyIORef' appStateRef $
+      swap . runState (runBaseWebCode contents)
+    return $ T.pack foo
```

Comment:
@byorgey, I got this to typecheck, but it doesn't seem to work.

To test this, I started a Classic game.  Executing `move;` in the REPL pane works. However, executing the following `curl` command has no effect on the game UI:

    curl -H "Content-Type: text/plain; charset=utf-8" -i http://localhost:5357/code/run --data "move;"

Any ideas?

Reply:
The `IORef` only stores a *copy* of the `AppState`, so modifying it is not going to affect the running game. See the code in https://github.com/swarm-game/swarm/blob/main/src/Swarm/App.hs and in particular how the `eventHandler` defined on line 87 always just reads the current game state and writes it into the `IORef`. 

Changing the `eventHandler` to somehow do two-way synchronization of the state sounds like a bad idea---since the web server and the game are running asynchronously in different threads, the state could get updated simultaneously in both places.  Probably what we should do instead is simply say that the copy of the `AppState` in the state monad is the single canonical state, and the `appStateRef` only ever contains a read-only copy.  If the web server wants to update the state it needs to send some appropriate event to the game's main event queue where it can be processed.  i.e. we can add a new `AppEvent` (see `Swarm.TUI.Model`) containing a `Text` value indicating that someone sent it to the `code/run` endpoint.

Reply:
Ah yes, that makes way more sense.  I'll admit it was satisfying to have the "type tetris" finally click into place with the `atomicModifyIORef` and `State` monad, so am mildly disappointed that it does absolutely nothing :smile: